feat(kad): implement automatic client mode (#3877)
paritytech/rust-libp2p: protocols/kad/CHANGELOG.md

- Remove deprecated public modules `handler`, `protocol` and `kbucket`.
  See [PR 3896].

- Automatically configure client/server mode based on external addresses.
  If we have or learn about an external address of our node, we operate in server-mode and thus allow inbound requests.

>   By default, all connections are in client-mode and only allow outbound requests.

Reviewer: ```suggestion
  By default, a node is in client-mode and only allows outbound requests.
```

Since the `Mode` is a global setting now.
